File: ProblemSolve/902.cs
/*************************************
 * Numbers At Most N Given Digit Set *
 *************************************/

//참고 : https://leetcode.com/problems/numbers-at-most-n-given-digit-set/solution/
public class Solution {
    public int AtMostNGivenDigitSet(string[] digits, int n) {
        string given = n.ToString();
        int len = given.Length;
        
        int[] dp = new int[len + 1];
        dp[len] = 1;
        
        for(int i=len-1; i>=0; --i){
            int givenDigit = given[i]-'0';
            foreach(string digit in digits){
                //더 작은 수이면 길이의 제곱 만큼 개수가 증가
                if(int.Parse(digit) < givenDigit){
                    dp[i] += (int)Math.Pow(digits.Length, len - i - 1);
                }
                //같은 수면 그대로 증가
                else if(int.Parse(digit) == givenDigit){
                    dp[i] += dp[i+1];
                }
            }
        }
        
        for(int i=1; i<len; ++i){
            dp[0] += (int)Math.Pow(digits.Length, i);
        }
        
        return dp[0];
    }
}

File: ProblemSolve/902.cpp
/*************************************
 * Numbers At Most N Given Digit Set *
 *************************************/

//참고 : https://leetcode.com/problems/numbers-at-most-n-given-digit-set/solution/

class Solution {
public:
    int atMostNGivenDigitSet(vector<string>& digits, int n) {
        string given = to_string(n);
        int len = given.size();
        int digitSize = digits.size();
        
        vector<int> dp(len + 1);
        dp[len] = 1;
        
        for(int i=len-1; i>=0; i--){
            int givensDigit = given[i] - '0';
            for(string digit : digits){
                if(stoi(digit) < givensDigit){
                    dp[i] += pow(digitSize, len - i - 1);
                }
                else if(stoi(digit) == givensDigit){
                    dp[i] += dp[i+1];
                }
            }
        }
        
        for(int i=1; i<len; i++){
            dp[0] += pow(digitSize, i);
        }
        
        return dp[0];
    }
};